### PR TITLE
MatrixBandPart operation for TensorFlow Hub models

### PR DESCRIPTION
**Pull request by himanshugupta11002**
### Details:
 - Implement the corresponding loader into op directory and named as matrix_band_part.cpp
    - according to hints given I was find the sizes m and n of two last dimension of input and generate ranges [0, m) and [0, n). 
      Unsqueeze these ranges so that to have tensor of shapes [m, 1] and [1, n]
 - Register it into the dictionary of loaders
 - Then apply the logical operations
 - Implement the layer test for this operation

Problem when I am running on Codespaces it give me so many errors on which I am currently working on.

### Tickets:
 - #22081

### CODE REVIEW

**Comment by rkazants** [comment]
build_jenkins

**Review by rkazants — COMMENTED** [review]
@himanshugupta11002, please update https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/tensorflow/docs/supported_ops.md documentation. It is recently added new check.

**Comment by himanshugupta11002** [comment]
okay I'll update it

**Comment by rkazants** [comment]
> okay I'll update it

Hi @himanshugupta11002, any update on the fix? It should be very easy:)
Please do not delay.

Best regards,
Roman

**Comment by rkazants** [comment]
build_jenkins

File: src/frontends/tensorflow_common/src/op/matrix_band_part.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#include "openvino/op/logical_and.hpp"
+#include "openvino/op/less_equal.hpp"
+#include "openvino/op/unsqueeze.hpp"
+#include "openvino/op/constant.hpp"
+#include "openvino/op/concat.hpp"
+#include "openvino/op/reshape.hpp"
+#include "utils.hpp"
+
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/matrix_band_part.cpp:8]: Please apply code formatting using clang

File: src/frontends/tensorflow_common/src/op/matrix_band_part.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+using namespace std;
+using namespace ov::op;
+
+namespace ov {
+namespace frontend {
+namespace tensorflow {
+namespace op {
+
+OutputVector matrix_band_part(const NodeContext& node) {
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/matrix_band_part.cpp:17]:
```suggestion
OutputVector translate_matrix_band_part_op(const NodeContext& node) {
```

File: src/frontends/tensorflow_common/src/op/matrix_band_part.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+#include "utils.hpp"
+
+using namespace std;
+using namespace ov::op;
+
+namespace ov {
+namespace frontend {
+namespace tensorflow {
+namespace op {
+
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/matrix_band_part.cpp:16]: please add declaration of `translate_matrix_band_part_op` in `openvino\src\frontends\tensorflow_common\include\common_op_table.hpp`

**Review by rkazants — COMMENTED** [review]
Please apply code formatting

**Comment by himanshugupta11002** [comment]
> Please apply code formatting

Okay, I'll try one more time

**Comment by rkazants** [comment]
> > Please apply code formatting
> 
> Okay, I'll try one more time

If you use IDE (like MSVS), you can try `Edit->Advanced->Format document` for each your source file. It should automatically adjust code format.

**Comment by rkazants** [comment]
Hi @himanshugupta11002, any update on resolving code-style? Please do not delay, code-style should be easy.

File: src/frontends/tensorflow_common/src/op/matrix_band_part.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
+#include "openvino/op/logical_and.hpp"
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/matrix_band_part.cpp:1]:
```suggestion
// Copyright (C) 2024 Intel Corporation
// SPDX-License-Identifier: Apache-2.0
//

#include "openvino/op/logical_and.hpp"
```

File: tests/layer_tests/tensorflow_tests/test_tf_MatrixBandPart.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import numpy as np
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_MatrixBandPart.py:1]:
```suggestion
# Copyright (C) 2018-2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import numpy as np
```

File: tests/layer_tests/tensorflow_tests/test_tf_MatrixBandPart.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        dict(input_shape=[2, 3, 5, 5]),
+    ]
+
+    @pytest.mark.parametrize("params", test_data_basic)
+    @pytest.mark.precommit_tf_fe
+    @pytest.mark.nightly
+    def test_matrix_band_part_basic(self, params, ie_device, precision, ir_version, temp_dir, use_new_frontend):
+        self._test(*self.create_matrix_band_part_net(**params),
+                   ie_device, precision, ir_version, temp_dir=temp_dir,
+                   use_new_frontend=use_new_frontend)
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_MatrixBandPart.py:35]: please avoid modifying `model_fp32.frozen` files below

**Review by rkazants — COMMENTED** [review]
please apply comment. code-formatting is not yet applied

**Comment by rkazants** [comment]
build_jenkins

**Comment by rkazants** [comment]
Hi @himanshugupta11002,

can you please apply comments asap because we have code freeze for 2024.0 release tomorrow and your PR is highly recommended to be included?

Best regards,
Roman

**Comment by rkazants** [comment]
build_jenkins

File: src/frontends/tensorflow_common/src/op/matrix_band_part.cpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    auto unsqueeze_range_n = make_shared<v0::Unsqueeze>(range_n, make_shared<v0::Constant>(element::i64, Shape{1}, {0}));
+
+    // Create indicator function using logical operations
+    auto in_band_rows = make_shared<v1::LogicalAnd>(make_shared<v1::LessEqual>(unsqueeze_range_m, m),
+                                                    make_shared<v1::GreaterEqual>(unsqueeze_range_m, make_shared<v0::Constant>(element::i64, Shape{}, {0})));
+
+    auto in_band_cols = make_shared<v1::LogicalAnd>(make_shared<v1::LessEqual>(unsqueeze_range_n, n),
+                                                    make_shared<v1::GreaterEqual>(unsqueeze_range_n, make_shared<v0::Constant>(element::i64, Shape{}, {0})));
+
+    auto in_band_indicator = make_shared<v1::LogicalAnd>(in_band_rows, in_band_cols);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/matrix_band_part.cpp:57]: you forget to use other inputs to this operation `num_lower` and `num_upper`. `in_band` computes as follows (see https://www.tensorflow.org/api_docs/python/tf/raw_ops/MatrixBandPart):

```
in_band(m, n) = (num_lower < 0 || (m-n) <= num_lower)) && (num_upper < 0 || (n-m) <= num_upper)
```

File: src/frontends/tensorflow_common/src/op/matrix_band_part.cpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+using namespace std;
+using namespace ov::op;
+
+namespace ov {
+namespace frontend {
+namespace tensorflow {
+namespace op {
+
+OutputVector translate_matrix_band_part_op(const NodeContext& node) {
+    // Input tensor and parameters
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/matrix_band_part.cpp:21]:
```suggestion
    default_op_checks(node, 2, {"AdjustContrastv2"});
    // Input tensor and parameters
```

File: tests/layer_tests/tensorflow_tests/test_tf_MatrixBandPart.py (hunk 6 above)
Comment (rkazants) [reply tests/layer_tests/tensorflow_tests/test_tf_MatrixBandPart.py:38]: please revert changes for `model_fp32.frozen` files

File: src/frontends/tensorflow_common/src/op/matrix_band_part.cpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+namespace ov {
+namespace frontend {
+namespace tensorflow {
+namespace op {
+
+OutputVector translate_matrix_band_part_op(const NodeContext& node) {
+    default_op_checks(node, 3, {"MatrixBandPart"});
+    // Input tensor and parameters
+    auto input = node.get_input(0);
+
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/matrix_band_part.cpp:24]: please get `num_lower` and `num_upper` as follows, since they are inputs according to the specification https://www.tensorflow.org/api_docs/python/tf/raw_ops/MatrixBandPart:

```
    auto num_lower = node.get_input(1);
    auto num_upper = node.get_input(1);
```

No need to create `Constant` nodes for them below

File: src/frontends/tensorflow_common/src/op/matrix_band_part.cpp [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    // Unsqueeze ranges to have tensors of shapes [m, 1] and [1, n]
+    auto unsqueeze_range_m = make_shared<v0::Unsqueeze>(range_m, make_shared<v0::Constant>(element::i64, Shape{1}, {1}));
+    auto unsqueeze_range_n = make_shared<v0::Unsqueeze>(range_n, make_shared<v0::Constant>(element::i64, Shape{1}, {0}));
+
+    // Create indicator function using logical operations
+    // Create indicator function using logical operations
+    auto in_band_rows = make_shared<v1::LogicalAnd>(
+        make_shared<v1::LogicalOr>(
+            make_shared<v1::Less>(make_shared<v1::Constant>(element::i64, Shape{}, {num_lower}), make_shared<v1::Constant>(element::i64, Shape{}, {0})),
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/matrix_band_part.cpp:58]: Here you create a constant for `num_lower`. This is not needed. Please pass `num_lower` as is. Also please create constant for zero by separate variable of the same type as `num_lower` and pass it to `LogicalOr`.
The same comment for `num_upper`

File: src/frontends/tensorflow_common/src/op/matrix_band_part.cpp [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        )
+    );
+
+    auto in_band_cols = make_shared<v1::LogicalAnd>(
+        make_shared<v1::LogicalOr>(
+            make_shared<v1::Less>(make_shared<v1::Constant>(element::i64, Shape{}, {num_lower}), make_shared<v1::Constant>(element::i64, Shape{}, {0})),
+            make_shared<v1::LessEqual>(make_shared<v1::Subtract>(m, n), make_shared<v1::Constant>(element::i64, Shape{}, {num_lower}))
+        ),
+        make_shared<v1::LogicalOr>(
+            make_shared<v1::Less>(make_shared<v1::Constant>(element::i64, Shape{}, {num_upper}), make_shared<v1::Constant>(element::i64, Shape{}, {0})),
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/matrix_band_part.cpp:75]: Please fix `num_lower` and `num_upper` usage here as well.

File: src/frontends/tensorflow_common/src/op/matrix_band_part.cpp [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+            make_shared<v1::LessEqual>(make_shared<v1::Subtract>(n, m), make_shared<v1::Constant>(element::i64, Shape{}, {num_upper}))
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/matrix_band_part.cpp:76]: num_upper

**Comment by rkazants** [comment]
Hi @himanshugupta11002, pls address comments above. I believe your PR is close to completeness.

Best regards,
Roman

**Comment by himanshugupta11002** [comment]
Hi @rkazants, should I add comments in the matrixbandpart.cpp file, or should I add them in all files where I have made changes?

File: tools/mo/unit_tests/moc_tf_fe/test_models/model_fp32.frozen
Comment (rkazants) [inline tools/mo/unit_tests/moc_tf_fe/test_models/model_fp32.frozen:1]: please revert this change

File: tools/ovc/unit_tests/moc_tf_fe/test_models/model_fp32.frozen
Comment (rkazants) [inline tools/ovc/unit_tests/moc_tf_fe/test_models/model_fp32.frozen:1]: please revert this change

File: tools/mo/unit_tests/moc_tf_fe/test_models/model_fp32.frozen
Comment (rkazants) [reply tools/mo/unit_tests/moc_tf_fe/test_models/model_fp32.frozen:1]: @himanshugupta11002, can you please revert this asap? That is because it is blocking to enable StableDiffusion model from Keras.

**Comment by rkazants** [comment]
Hi @himanshugupta11002, I am closing this PR. I fix your implementation there: https://github.com/openvinotoolkit/openvino/pull/23082. Your contribution is preserved there.
I did so because we urgently need this translator for Keras StableDiffusion model support.

Best regards,
Roman
